File: 135.分发糖果.cpp
/*
 * @lc app=leetcode.cn id=135 lang=cpp
 *
 * [135] 分发糖果
 */

// @lc code=start
class Solution {
public:
    int candy(vector<int>& ratings) {
        vector<int> res(ratings.size(), 1);
        for (int i = 1; i < ratings.size(); ++i) {
            if (ratings[i] > ratings[i-1]) {
                res[i] = res[i-1] + 1;
            }
        }
        for (int i = ratings.size() - 2; i >= 0; --i) {
            if (ratings[i] > ratings[i+1]) {
                res[i] = max(res[i+1]+1, res[i]);
            }
        }
        return accumulate(res.begin(), res.end(), 0);
    }
};
// @lc code=end

File: 135.分发糖果.py
#
# @lc app=leetcode.cn id=135 lang=python
#
# [135] 分发糖果
#

# 从左到右遍历一遍，如果右边比左边的分数大，就以左边的+1
# 从右往左遍历一遍，如果左边比右边分数大，取 左边当前值和右边值+1 两者中的max
# @lc code=start
class Solution(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        res = [1] * len(ratings)
        for i in range(1, len(ratings)):
            if ratings[i] > ratings[i - 1]:
                res[i] = res[i - 1] + 1
        for i in reversed(range(len(ratings) - 1)):
            if ratings[i] > ratings[i + 1]:
                res[i] = max(res[i], res[i + 1] + 1)
        return sum(res)
# @lc code=end